surveyor/jetstream.go

```diff
@@ -0,0 +1,199 @@
+// Copyright 2020 The NATS Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package surveyor
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/nats-io/jsm.go"
+	"github.com/nats-io/jsm.go/api"
+	"github.com/nats-io/nats.go"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// JSAdvisoryListener listens for JetStream advisories and expose them as prometheus data
+type JSAdvisoryListener struct {
+	nc    *nats.Conn
+	opts  *jsAdvisoryOptions
+	sopts *Options
+}
+
+type jsAdvisoryOptions struct {
+	AccountName string `json:"name"`
+	Credentials string `json:"credentials"`
+}
+
+// Validate checks the options meet our expectations
+func (o *jsAdvisoryOptions) Validate() error {
+	errs := []string{}
+
+	if o.AccountName == "" {
+		errs = append(errs, "name is required")
+	}
+
+	if o.Credentials != "" {
+		_, err := os.Stat(o.Credentials)
+		if err != nil {
+			errs = append(errs, fmt.Sprintf("invalid credential file: %s", err))
+		}
+	}
+
+	if len(errs) == 0 {
+		return nil
+	}
+
+	return fmt.Errorf(strings.Join(errs, ", "))
+}
+
+var (
+	// API Audit
+	jsAPIAuditCtr = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name: prometheus.BuildFQName("nats", "survey", "jetstream_api_audit"),
+		Help: "JetStream API access audit events",
+	}, []string{"server", "subject", "account"})
+
+	jsAPIErrorsCtr = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name: prometheus.BuildFQName("nats", "survey", "jetstream_api_errors"),
+		Help: "JetStream API Errors Count",
+	}, []string{"server", "subject", "account"})
+
+	// Delivery Exceeded
+	jsDeliveryExceededCtr = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name: prometheus.BuildFQName("nats", "survey", "jetstream_delivery_exceeded_count"),
+		Help: "Advisories about JetStream Consumer Delivery Exceeded events",
+	}, []string{"account", "stream", "consumer"})
+
+	// Ack Samples
+	jsAckMetricDelay = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+		Name: prometheus.BuildFQName("nats", "survey", "jetstream_acknowledgement_duration"),
+		Help: "How long an Acknowledged message took to be Acknowledged",
+	}, []string{"account", "stream", "consumer"})
+
+	jsAckMetricDeliveries = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name: prometheus.BuildFQName("nats", "survey", "jetstream_acknowledgement_deliveries"),
```

Comment:
Per prometheus recommendations, should this be `jetstream_acknowedgement_delivery_count`?

Reply:
yup, will fix